### PR TITLE
Issue 2380 and 2388

### PR DESCRIPTION
**Pull request by dnj12345**
Addresses issues [2388](https://github.com/apache/trafficserver/issues/2388)  and [2380](https://github.com/apache/trafficserver/issues/2380)

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci].

**Comment by zwoop** [comment]
Also, was this discussed / vetted on the mailing list? Not sure I'm seeing it.

**Comment by zwoop** [comment]
I'm wondering too if it should just be TSVConnDataSet()/Get(), which is inline with the existing TSContDataSet()/Get().

**Comment by dnj12345** [comment]
Hi Leif,
This is my first time making a submission and hence didn’t know the process. I thought the documentation was reviewed along with the pr. Anyways, I’ll make the changes to combine the docs and remember to get approval first. Thanks 

Bhasker.

**Comment by maskit** [comment]
@dnj12345 I think Leif is talking about [API review process](https://cwiki.apache.org/confluence/display/TS/API+Review+Process). You don't need a prior approval  for submitting a PR that adds new APIs, however, you need to send a API proposal to the mailing list.

Also, It seems like #2380 and #2388 are not related to / depend on each other to me. I understand you have an example plugin and it uses the both features but maybe you should send 2 or 3 separated PRs.

**Comment by dnj12345** [comment]
Understood about the api review process. However, as I mentioned, this is
my first time
and I didn't know the process.

About combining 2380 and 2388, since our company module needed resolution
for both
these issues, I created the PR handling both issues. Since I'd like to back
port this fix to
older releases I'd like to keep them together. Otherwise, I'll have to
submit a multitude of PRs.

**Comment by dnj12345** [comment]
ok. Can I go ahead and make this change?

**Comment by maskit** [comment]
Ah, now I found your emails. It's "ssl conn and http", right? It may not get enough attention. I recommend you to start another thread with a subject like "[API PROPOSAL] ..." as an official proposal.

File: iocore/net/I_NetVConnection.h [hunk 1]
```diff
@@ -658,6 +667,17 @@ class NetVConnection : public VConnection
   int write_buffer_empty_event;
   /// NetVConnection Context.
   NetVConnectionContext_t netvc_context;
+
+private:
+  struct user_data {
+    uint8_t name_len;
```
Comment (SolidWallOfCode) [inline iocore/net/I_NetVConnection.h]: Use `ts::string_view`, not an explicit pointer / length.

**Comment by zwoop** [comment]
I agree with @SolidWallOfCode , we should implement this as per the discussions in #2388. That would be much more consistent with the existing APIs for managing data like this, and is important to allow for multiple plugins to safely use this type of data on the VC.

**Comment by dnj12345** [comment]
Alan,
My initial thought process was similar to yours. However, on examining the
code, it looked like
the ts_vconn close was getting called from the higher layers after they had
cleaned up their
data-structures. Since I still needed access to the ssl data-structures and
since there was
an event framework already existing for ssl, I took this route.

**Comment by dnj12345** [comment]
Hi Leif, please see my response to alan, I sent a few minutes. I understand
there is a need for
close hook at the vconn level, however, I wouldn't solve my issue where i
need access to the
higher layer data structures...

**Comment by dnj12345** [comment]
will do. thanks.

**Comment by SolidWallOfCode** [comment]
I wonder if we want both a `TS_SSL_CLOSE` and `TS_VCONN_CLOSE`. We already have precedent for TXN vs. SSN close, it might be reasonable to have a close event for each protocol layer. The original reason for having `TS_VCONN_CLOSE` was for connections that didn't make it up to the higher layers. It seems possible that could happen at the TLS layer as well.

**Comment by SolidWallOfCode** [comment]
We've split the TXN and SSN args as recommended to make this fit well in to the scheme of hooks. I think it's time to update the PR to use that style of index reservation.

**Comment by dnj12345** [comment]
Alan, Could you point me to the PR where these changes were made. Thanks.

**Comment by dnj12345** [comment]
Alan,

"Generally this will be a file or library scope global which is set at
plugin initialization. Note the reservation is by convention - nothing
stops a plugin from interacting with a TSVConn arg it has not reserved."

Is there any reason for requiring this to be in global scope and not
connection scope?
If the plugin needs to save connection specific data, global scope may not
work since
we may have lots of connections.

**Comment by SolidWallOfCode** [comment]
@dnj12345 the changes are in #2905 - that's noted in the comment immediately before the one you quoted.

There's nothing specifically that requires it to be in the global scope, that's just generally how it's used. However, as there is no mechanism for releasing a reserved argument, tracking the reserved index in a global or file scope value is the best choice.

> If the plugin needs to save connection specific data, global scope may not work since we may have lots of connections.

The arguments *slots* are per transaction (for TXN args) and per session (for SSN args). The slot *index* is what is reserved. Therefore using the same index for every transaction for a TXN arg works without collisions. E.g., if a plugin reserves index `2`, then it can use the TXN slot `2` in every transaction independently of other plugins and between transactions.

**Comment by dnj12345** [comment]
Closing it so that I can redo it based on the comments.

**Pull request closed without merging**